### PR TITLE
code refactoring: cost effective gradient boosting

### PR DESCRIPTION
**Pull request by guolinke**
refer to #2341

### CODE REVIEW

File: src/treelearner/serial_tree_learner.cpp [hunk 1]
```diff
@@ -69,8 +70,6 @@ void SerialTreeLearner::Init(const Dataset* train_data, bool is_constant_hessian
   histogram_pool_.DynamicChangeSize(train_data_, config_, max_cache_size, config_->num_leaves);
   // push split information for all leaves
   best_split_per_leaf_.resize(config_->num_leaves);
-  splits_per_leaf_.resize(config_->num_leaves*train_data_->num_features());
```
Comment (guolinke) [inline src/treelearner/serial_tree_learner.cpp]: @StrikerRUS Here is an additional cost, before this PR. So I think this should be in v2.3.0 release as well

Comment (StrikerRUS) [reply src/treelearner/serial_tree_learner.cpp]: @guolinke Sure! We have been waiting so long, so can wait a little bit more, I guess.

**Comment by guolinke** [comment]
@StrikerRUS any style issues?

**Comment by StrikerRUS** [comment]
@guolinke 
> any style issues?

I'll run `cpplint` on this branch tomorrow.

**Comment by guolinke** [comment]
@StrikerRUS maybe we can add cpplint into CI, but don't throw the error? (or ignore some types?)

**Comment by StrikerRUS** [comment]
@guolinke 
> maybe we can add cpplint into CI, but don't throw the error? (or ignore some types?)

I like this idea! I'll create a PR for that.

> any style issues?

Here is a part of report from `cpplint` related to files with diff:
```
./src\treelearner\serial_tree_learner.h:5:  #ifndef header guard has wrong style, please use: SRC_TREELEARNER_SERIAL_TREE_LEARNER_H_  [build/header_guard] [5]
./src\treelearner\serial_tree_learner.h:195:  #endif line should be "#endif  // SRC_TREELEARNER_SERIAL_TREE_LEARNER_H_"  [build/header_guard] [5]
./src\treelearner\serial_tree_learner.h:32:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
./src\treelearner\serial_tree_learner.h:83:  Do not leave a blank line after "protected:"  [whitespace/blank_line] [3]
./src\treelearner\serial_tree_learner.h:111:  Is this a non-const reference? If so, make const or use a pointer: Json& forced_split_json  [runtime/references] [2]
Done processing ./src\treelearner\serial_tree_learner.h

./src\treelearner\serial_tree_learner.cpp:5:  Include the directory when naming .h files  [build/include_subdir] [4]
./src\treelearner\serial_tree_learner.cpp:8:  Found C system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:9:  Found C system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:10:  Found C system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:11:  Found C system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:13:  Found C++ system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:14:  Found C++ system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:15:  Found C++ system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:16:  Found C++ system header after other header. Should be: serial_tree_learner.h, c system, c++ system, other.  [build/include_order] [4]
./src\treelearner\serial_tree_learner.cpp:288:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
./src\treelearner\serial_tree_learner.cpp:291:  Missing space before ( in if(  [whitespace/parens] [5]
./src\treelearner\serial_tree_learner.cpp:300:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
./src\treelearner\serial_tree_learner.cpp:312:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
./src\treelearner\serial_tree_learner.cpp:874:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
Done processing ./src\treelearner\serial_tree_learner.cpp

./src\treelearner\cost_effective_gradient_boosting.hpp:0:  No copyright message found.  You should have a line: "Copyright [year] <Copyright Owner>"  [legal/copyright] [5]
./src\treelearner\cost_effective_gradient_boosting.hpp:0:  No #ifndef header guard found, suggested CPP variable is: SRC_TREELEARNER_COST_EFFECTIVE_GRADIENT_BOOSTING_HPP_  [build/header_guard] [5]
./src\treelearner\cost_effective_gradient_boosting.hpp:7:  Include the directory when naming .h files  [build/include_subdir] [4]
./src\treelearner\cost_effective_gradient_boosting.hpp:13:  public: should be indented +1 space inside class CostEfficientGradientBoosting  [whitespace/indent] [3]
./src\treelearner\cost_effective_gradient_boosting.hpp:14:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
./src\treelearner\cost_effective_gradient_boosting.hpp:15:  Redundant blank line at the start of a code block should be deleted.  [whitespace/blank_line] [2]
./src\treelearner\cost_effective_gradient_boosting.hpp:15:  Redundant blank line at the end of a code block should be deleted.  [whitespace/blank_line] [3]
./src\treelearner\cost_effective_gradient_boosting.hpp:75:  private: should be indented +1 space inside class CostEfficientGradientBoosting  [whitespace/indent] [3]
./src\treelearner\cost_effective_gradient_boosting.hpp:76:  Do not leave a blank line after "private:"  [whitespace/blank_line] [3]
./src\treelearner\cost_effective_gradient_boosting.hpp:101:  Add #include <vector> for vector<>  [build/include_what_you_use] [4]
Done processing ./src\treelearner\cost_effective_gradient_boosting.hpp
```

I guess, a fix for all headers order issues will be placing `#include "cost_effective_gradient_boosting.hpp"` after `#include <utility>` with one blank line. I remember we have [the following order of includes](https://github.com/microsoft/LightGBM/pull/2066#issue-263709049):

```
dir2/foo2.h (or "foo2.hpp")
A blank line
Your project's .h files in include folder  (in the alphabetical order).
A blank line
C system files (in the alphabetical order).
C++ system files (in the alphabetical order).
A blank line
Other libraries' .h (or .hpp) files  (in the alphabetical order).
Your project's .h (or .hpp) files in src folder  (in the alphabetical order).
```

https://github.com/microsoft/LightGBM/blob/023749239ebf390b5d56e08f227a013cbd972eb0/src/io/parser.cpp#L5-L11
https://github.com/microsoft/LightGBM/blob/023749239ebf390b5d56e08f227a013cbd972eb0/src/io/bin.cpp#L5-L18

**Comment by StrikerRUS** [comment]
Remaining cpplint errors:
```
./src\treelearner\cost_effective_gradient_boosting.hpp:16:  Include the directory when naming .h files  [build/include_subdir] [4]
./src\treelearner\cost_effective_gradient_boosting.hpp:82:  private: should be indented +1 space inside class CostEfficientGradientBoosting  [whitespace/indent] [3]
./src\treelearner\cost_effective_gradient_boosting.hpp:82:  "private:" should be preceded by a blank line  [whitespace/blank_line] [3]
./src\treelearner\cost_effective_gradient_boosting.hpp:111:  At least two spaces is best between code and comments  [whitespace/comments] [2]
Done processing ./src\treelearner\cost_effective_gradient_boosting.hpp

./src\treelearner\serial_tree_learner.cpp:5:  Include the directory when naming .h files  [build/include_subdir] [4]
./src\treelearner\serial_tree_learner.cpp:289:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
./src\treelearner\serial_tree_learner.cpp:292:  Missing space before ( in if(  [whitespace/parens] [5]
./src\treelearner\serial_tree_learner.cpp:301:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
./src\treelearner\serial_tree_learner.cpp:313:  Consider using CHECK_GE instead of CHECK(a >= b)  [readability/check] [2]
./src\treelearner\serial_tree_learner.cpp:875:  Consider using CHECK_GT instead of CHECK(a > b)  [readability/check] [2]
Done processing ./src\treelearner\serial_tree_learner.cpp

./src\treelearner\serial_tree_learner.h:32:  Do not use namespace using-directives.  Use using-declarations instead.  [build/namespaces] [5]
./src\treelearner\serial_tree_learner.h:83:  Do not leave a blank line after "protected:"  [whitespace/blank_line] [3]
./src\treelearner\serial_tree_learner.h:111:  Is this a non-const reference? If so, make const or use a pointer: Json& forced_split_json  [runtime/references] [2]
Done processing ./src\treelearner\serial_tree_learner.h
```

File: src/treelearner/cost_effective_gradient_boosting.hpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
+/*!
+ * Copyright (c) 2016 Microsoft Corporation. All rights reserved.
```
Comment (StrikerRUS) [inline src/treelearner/cost_effective_gradient_boosting.hpp]: I guess it should be `2019`. Refer to https://github.com/microsoft/LightGBM/pull/2101#issue-270098443.

File: src/treelearner/cost_effective_gradient_boosting.hpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  }
+  void Init() {
+    auto train_data = tree_learner_->train_data_;
+    splits_per_leaf_.resize(static_cast<size_t>(tree_learner_->config_->num_leaves) * train_data->num_features());
+    is_feature_used_in_split_.clear();
+    is_feature_used_in_split_.resize(train_data->num_features());
+
+    if (!tree_learner_->config_->cegb_penalty_feature_coupled.empty()) {
+      CHECK(tree_learner_->config_->cegb_penalty_feature_coupled.size() == static_cast<size_t>(train_data->num_total_features()));
```
Comment (StrikerRUS) [inline src/treelearner/cost_effective_gradient_boosting.hpp]: Maybe `Log:Fatal` would be better, as it's check for user input?

File: src/treelearner/cost_effective_gradient_boosting.hpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    if (!tree_learner_->config_->cegb_penalty_feature_lazy.empty()) {
+      CHECK(tree_learner_->config_->cegb_penalty_feature_lazy.size() == static_cast<size_t>(train_data->num_total_features()));
```
Comment (StrikerRUS) [inline src/treelearner/cost_effective_gradient_boosting.hpp]: Same as above about `Log:Fatal`.

File: src/treelearner/cost_effective_gradient_boosting.hpp [hunk 5]
```diff
@@ -107,5 +108,5 @@ class CostEfficientGradientBoosting {
   std::vector<uint32_t> feature_used_in_data_;
 };
 
-}  // namespace LightGBM
```
Comment (StrikerRUS) [inline src/treelearner/cost_effective_gradient_boosting.hpp]: Wrong line number due to added blank line. Original cpplint issue was about `#endif` line (112 now).

**Comment by guolinke** [comment]
Thanks @StrikerRUS , I a

**Comment by StrikerRUS** [comment]
I guess, https://github.com/microsoft/LightGBM/blob/master/windows/LightGBM.vcxproj should be updated as well.

**Comment by guolinke** [comment]
@StrikerRUS any ideas about the python lint error?

**Comment by StrikerRUS** [comment]
@guolinke Errors are due to the new version of `pydocstyle`, refer to #2451. I introduced new label `blocking`, which means that this PR should be merged firstly to pass CI (or maybe another reason).

**Pull request merged**